### PR TITLE
ENH Add plot_roc_curve_by_group for ROC curves by sensitive feature

### LINKED ISSUE #758: plot_roc_curves(scores, y_true, sensitive_features)
#### Is your feature request related to a problem? Please describe.
In producing Fairness Assessment results, seeing metrics given a fixed threshold is useful. In addition to that, it helps to also see an ROC curve (TPR vs. FPR) across different thresholds. 

#### Describe the solution you'd like
A function like "plot_roc_curves(scores, y_true, sensitive_features)" that takes in sensitive_features (list of sensitive features) as a parameter, and plots roc curves across subgroups

#### Describe alternatives you've considered, if relevant
sklearn plot_roc_curve, though this does not support plot via sensitive groups 

#### Additional context
Creating a Fairness Assessment using Fairlearn, ROC curve plots are missing

### PR DESCRIPTION
**Pull request by kstohr**
## Summary

Adds `fairlearn.metrics.plot_roc_curve_by_group`, a function that plots ROC
curves disaggregated by sensitive feature so you can see whether a binary
classifier separates the positive and negative classes equally well across
groups. For each subgroup it draws one ROC curve (with its AUC in the legend),
plus the overall curve and a chance-level baseline for reference.

```python
from fairlearn.metrics import plot_roc_curve_by_group

ax = plot_roc_curve_by_group(
    y_true,
    y_score,                      # scores for the positive class
    sensitive_features=A_test,    # one or more sensitive features
)
```

When multiple sensitive features are passed, their value combinations define the
subgroups (e.g. `"Female,White"`), matching the `MetricFrame` convention.

## What changed

- **New public API** `plot_roc_curve_by_group(y_true, y_score, *, sensitive_features, ax=None, plot_overall=True, plot_chance_level=True, pos_label=None, title=None) -> matplotlib.axes.Axes`.
- This PR was stale, so the branch was first **brought up to date with `main`** (which had since refactored `fairlearn.metrics`).
- The original `RocAuc` class was **redesigned into a function** following Fairlearn's plotting conventions (cf. `plot_model_comparison`). Group splitting now uses Fairlearn's input validation plus scikit-learn's `RocCurveDisplay.from_predictions`, instead of misusing `MetricFrame` with a dummy metric. AUC scores are obtained the idiomatic way, via `MetricFrame(metrics=roc_auc_score, ...)` (shown in the example).
- **matplotlib stays optional**: importing `fairlearn.metrics` no longer imports matplotlib; the plotting function raises the standard `RuntimeError` if matplotlib is missing.
- Supports numeric **and non-numeric** class labels (with `pos_label`), and `list` / `numpy` / `pandas` / `dict` sensitive-feature inputs.

## Docs & tests

- New worked example `examples/plot_roc_auc.py` (diabetes-hospital dataset), wired into the **Plotting** user guide.
- API-reference entry and a `v0.15.0` changelog entry.
- New unit tests in `test/unit/metrics/test_plot_roc_curve_by_group.py` (curve counts, legend/AUC correctness against scikit-learn, input-type handling, `pos_label`, overall-only / chance-only) plus a no-matplotlib regression test.

### CODE REVIEW

File: fairlearn/metrics/_roc_auc.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    Uses the standard Scikit-learn modules `roc_curve`, `auc`, `roc_auc_score`
+    to compute the false positive rate, true positive rate and area under the
+    curve (AUC) respectively. Passing keyword arguemnts to class methods will
+    pass those parameters to the underlying Scikit-learn function to allow for
+    greater control.
+
+    Reference
+    ----------
+    MetricFrame:
+        https://fairlearn.org/v0.6.2/api_reference/fairlearn.metrics.html
```
Comment (riedgar-ms) [inline fairlearn/metrics/_roc_auc.py:25]: This should be a reference to the class in the docs, not a full URL

File: fairlearn/metrics/_roc_auc.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+        # Plot overall
+        if include_overall:
+            ax = self.plot_overall(ax=ax)
+
+        # Plot ROC Curves by group
+        for name in sensitive_index:
+            grp_y_true = self.sensitive_series[name][0]
+            grp_y_score = self.sensitive_series[name][1]
+            display = self.plot_roc(
```
Comment (riedgar-ms) [inline fairlearn/metrics/_roc_auc.py:290]: Is it possible to use `self.plot_roc` as the metric itself, with judicious use of `functools.partial` to bind the other arguments (not sure if that works with `**kwargs`).

**Review by romanlutz — COMMENTED** [review]
@kstohr thanks for sharing, this is great! I've put in a few preliminary comments. It would be nice to have a few screenshots of how it renders in the example, and a brief PR description that links to the corresponding issue. Let me know if you have questions or doubts about my comments. I'm always happy to chat!

File: examples/plot_roc_auc.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+"""
+====================================
+Plot ROC Curves by Sensitive Feature
+====================================
+"""
+# %%
+# # Plot ROC Curve by Sensitive Feature
+
+# To understand the performance of a classification model, it is common to plot # the Receiver Operator Curve (ROC). The ROC Curve helps to visualize the trade # off between between sensitivity (or True Positive Rate) and specificity (1 – # False Positive Rate) for all possible thresholds of a classification model.
```
[the remark this answers is not in the document]
Comment (kstohr) [reply examples/plot_roc_auc.py:12]: Yep. will fix.

File: fairlearn/metrics/_roc_auc.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+from sklearn.metrics import roc_curve, auc, roc_auc_score
+from sklearn.metrics import RocCurveDisplay
+
+from fairlearn.metrics import MetricFrame
+
+class RocAuc:
+    """
+    Provides utilties for generating AUC scores, ROC curves
+    and plotting ROC curves grouped by sensitive feature. Depends on
```
[the remark this answers is not in the document]
Comment (kstohr) [reply fairlearn/metrics/_roc_auc.py:14]: Yep. "disaggregated based on one or more sensitive features" works for me.

File: fairlearn/metrics/_roc_auc.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+        self.y_true = y_true
+        self.y_score = y_score
+        self.sensitive_features = sensitive_features
+        self.sensitive_series = self.by_group()
+        self.overall_auc = roc_auc_score(self.y_true, self.y_score)
+        self.auc_scores = {}
+        self.__display = None
+
+    @staticmethod
+    def splitter(y_true, y_pred):
```
[the remark this answers is not in the document]
Comment (kstohr) [reply fairlearn/metrics/_roc_auc.py:78]: Yeah, especially this method. It does nothing.

**Comment by kstohr** [comment]
Ok, let me make those changes. We still have to build tests and I didn't actually run sphinx. I wrote the example, but didn't really look at how it will format in the docs. I also would like to give some direction on how you would mitigate a model once having run the ROC Curves.

File: fairlearn/metrics/_roc_auc.py (hunk 2 above)
Comment (kstohr) [reply fairlearn/metrics/_roc_auc.py:290]: @riedgar-ms  Yeah, I tried that at first. It may be possible, but when I tried it I ran into some issues with the MetricFrame class because it computes the metric on instantiation, and all I am actually using from that class is the group by function to split the data. It was far easier to pass a placeholder function. Open to suggestions. However, I think the work that needs to happen is some small tweaks to MetricFrame..So holding out for us to determine the approach to MetricFrame and then we can come back to this and refactor it if needed.

File: fairlearn/metrics/_roc_auc.py (hunk 1 above)
Comment (kstohr) [reply fairlearn/metrics/_roc_auc.py:25]: Yep. Will fix when I run the documentation (I didn't run Sphinx yet.)

File: fairlearn/metrics/_roc_auc.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+from sklearn.metrics import roc_curve, auc, roc_auc_score
+from sklearn.metrics import RocCurveDisplay
+
+from fairlearn.metrics import MetricFrame
+
+class RocAuc:
+    """
+    Provides utilties for generating AUC scores, ROC curves
+    and plotting ROC curves grouped by sensitive feature. Depends on
+    the Fairlearn MetricFrame class to split the input by sensitive feature.
```
[the remark this answers is not in the document]
Comment (kstohr) [reply fairlearn/metrics/_roc_auc.py:15]: @romanlutz Will that markdown automatically create a link to the API in the docs when I build the docs?

File: fairlearn/metrics/_roc_auc.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+    Uses the standard Scikit-learn modules `roc_curve`, `auc`, `roc_auc_score`
```
[the remark this answers is not in the document]
Comment (kstohr) [reply fairlearn/metrics/_roc_auc.py:16]: Yep... Can be more precise in my description. Def.

File: fairlearn/metrics/_roc_auc.py [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+    roc_auc_score:
+        https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score
+    RocCurveDisplay:
+        https://scikit-learn.org/stable/modules/generated/sklearn.metrics.RocCurveDisplay.html#sklearn.metrics.RocCurveDisplay
+
+
+    Parameters
+    ----------
+    y_true : List, pandas.Series, numpy.ndarray, pandas.DataFrame
+    The ground-truth labels for classification. (i.e. results of clf.predict())
```
[the remark this answers is not in the document]
Comment (kstohr) [reply fairlearn/metrics/_roc_auc.py:37]: Yeah. I didn't run the docs and test the formatting etc. I'll work through all of that...

File: examples/plot_roc_auc.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+# feature subgroups.
+#
+# One way to tune the model, is to adjust the threshold used by the decision
+# function. In this way, we can make the model more or less sensitive. This may # result in a model that is less accurate overall, but performs more "fairly."
+#
+# The RocAuc module in Fairlearn allows you to easily plot ROC Curves and
+# compute AUC scores by sensitive feature. Below is an example of how to use
+# the module and customize the resulting plots.
+#
+# **Note: To use this module, you will need to install Matplotlib which is not installed as part of Fairlearn**
```
[the remark this answers is not in the document]
Comment (kstohr) [reply examples/plot_roc_auc.py:42]: "To install Matplotlib and other custom plotting dependencies, install Fairlearn with the `customplots` extras: `fairlearn[customplots]`

File: examples/plot_roc_auc.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+# In visualizing the sensitive features by subgroup, a model that resulted inthe
+# same ROC Curve for all subgroups could be considered a perfectly "fair"
+# model. (In the real world, of course, it is not that simple and other factors
+# come into play.) It was able to determine the class for each subgroup at each
+# threshold equally well. If there is variation between the ROC curves, it is
+# an indication that the model performs differently for different sensiive
+# feature subgroups.
+#
+# One way to tune the model, is to adjust the threshold used by the decision
+# function. In this way, we can make the model more or less sensitive. This may # result in a model that is less accurate overall, but performs more "fairly."
```
[the remark this answers is not in the document]
Comment (kstohr) [reply examples/plot_roc_auc.py:36]: Yep. I'll make that change.

File: fairlearn/metrics/_roc_auc.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (romanlutz) [reply fairlearn/metrics/_roc_auc.py:15]: Yes! It's restructured text (ReST) rather than markdown, but that's exactly what it does. I go back to the ReST guide quite a lot these days: https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html

**Comment by romanlutz** [comment]
@kstohr thanks for patiently working through the comments 🙂 

In terms of mitigation I would suggest restricting this PR to generating the ROC curves. It will keep growing in scope otherwise and there may be different viewpoints on how one can mitigate that, which will drag everything out quite a bit. It's probably preferable to keep the change as small as possible and then add to it in follow-up PRs. We can certainly get the conversation started, though!

**Comment by romanlutz** [comment]
> @kstohr do you want to revisit this and perhaps set up some time to talk about testing as we had planned a while back?

Ping @kstohr 🙂 I suppose we can go ahead an assume that others can take over this PR if we don't hear back by 12/20 (two weeks from now). If you're still interested in getting back to this at a later point, @kstohr, feel free to say so!

**Comment by kstohr** [comment]
Sure thing. Let’s find a time.

**Review by romanlutz — APPROVED** [review]

**Review by riedgar-ms — APPROVED** [review]
No objections from me
